WalletWasabi/BitcoinCore/Mempool/MempoolMirror.cs

```diff
@@ -0,0 +1,142 @@
+		private object MempoolLock { get; } = new();
+
+		public override Task StartAsync(CancellationToken cancellationToken)
+		{
+			Node.MempoolService.TransactionReceived += MempoolService_TransactionReceived;
+			return base.StartAsync(cancellationToken);
+		}
+
+		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+		{
+			var sw = Stopwatch.StartNew();
+			int added = await MirrorMempoolAsync(stoppingToken).ConfigureAwait(false);
+			Logger.LogInfo($"{added} transactions were copied from the full node to the in-memory mempool within {sw.Elapsed.TotalSeconds} seconds.");
+
+			await base.ExecuteAsync(stoppingToken).ConfigureAwait(false);
+		}
+
+		public override Task StopAsync(CancellationToken cancellationToken)
+		{
+			Node.MempoolService.TransactionReceived -= MempoolService_TransactionReceived;
+			return base.StopAsync(cancellationToken);
+		}
+
+		private void MempoolService_TransactionReceived(object? sender, SmartTransaction stx)
+		{
+			AddTransactions(stx.Transaction);
+		}
+
+		private int AddTransactions(params Transaction[] txs) => AddTransactions(txs as IEnumerable<Transaction>);
+
+		private int AddTransactions(IEnumerable<Transaction> txs)
+		{
+			var added = 0;
+			lock (MempoolLock)
+			{
+				foreach (var tx in txs.Where(x => !Mempool.ContainsKey(x.GetHash())))
+				{
+					// Evict double spents.
+					EvictSpendersNoLock(tx.Inputs.Select(x => x.PrevOut));
+
+					Mempool.Add(tx.GetHash(), tx);
+					added++;
+				}
+			}
+			return added;
+		}
+
+		private void EvictSpendersNoLock(IEnumerable<OutPoint> txOuts)
+		{
+			HashSet<uint256> doubleSpents = new();
+			foreach (var input in txOuts)
+			{
+				foreach (var mempoolTx in Mempool)
+				{
+					if (mempoolTx.Value.Inputs.Select(x => x.PrevOut).Contains(input))
+					{
+						doubleSpents.Add(mempoolTx.Key);
+					}
+				}
+			}
+
+			foreach (var txid in doubleSpents)
+			{
+				Mempool.Remove(txid);
+			}
+		}
+
+		protected override async Task ActionAsync(CancellationToken cancel)
+		{
+			await MirrorMempoolAsync(cancel).ConfigureAwait(false);
+		}
+
+		private async Task<int> MirrorMempoolAsync(CancellationToken cancel)
+		{
+			var mempoolHashes = await Rpc.GetRawMempoolAsync().ConfigureAwait(false);
+
+			IEnumerable<uint256> missing;
+			lock (MempoolLock)
+			{
+				missing = mempoolHashes.Except(Mempool.Keys);
+
+				foreach (var txid in Mempool.Keys.Except(mempoolHashes).ToHashSet())
+				{
+					Mempool.Remove(txid);
+				}
+			}
+
+			var added = AddTransactions(Node.GetMempoolTransactions(missing, cancel));
+			return added;
+		}
+
+		public IEnumerable<Transaction> GetSpenderTransactions(IEnumerable<OutPoint> txOuts)
+		{
+			Dictionary<uint256, Transaction> spenders = new();
+			lock (MempoolLock)
+			{
+				foreach (var input in txOuts)
+				{
+					foreach (var mempoolTx in Mempool)
+					{
+						if (mempoolTx.Value.Inputs.Select(x => x.PrevOut).Contains(input))
```

Comment:
Like before, you can use HashSet here to speed this up.

Reply:
There's no speed bottleneck, but there's a readability bottleneck here. Adding a `HashSet` there is like Elon Musk using a coupon.

Reply:
I am not sure about that, MirrorMempoolAsync runs periodically. Every new tx EvictSpendersNoLock will run on the whole Mempool (~50000 tx) on all input of the new tx. 

So for example if the new tx is a CJ then 200 input * (50000tx's every inputs). Maybe there won't be a lot new txs later but every time you have to iterate through the while mempool all tx all inputs. 

Can you make some measurements on real data?

Reply:
No, sorry, I don't have time for bikeshedding now. If you still want to do it, then do it yourself.

Reply:
Right, but the proof of burden is on you. There's no performance issue until proven otherwise. Also, check out [Premature Optimization Is the Root of All Evil](https://stackify.com/premature-optimization-evil/). article.

Reply:
Though my point about bikeshedding is more relevant.

Reply:
Alright, we can try it.